test/EntityManagerFactoryTest.php

```diff
@@ -75,25 +79,38 @@ public function testConfigKeyTakenFromConfig() : void
                     ],
                 ],
             ],
-        ]);
-        $container->has('doctrine.connection.orm_foo')->willReturn(true);
-        $container->get('doctrine.connection.orm_foo')->willReturn($connection);
-        $container->has('doctrine.configuration.orm_bar')->willReturn(true);
-        $container->get('doctrine.configuration.orm_bar')->willReturn($configuration);
+        ];
+
+        $container = $this->createMock(ContainerInterface::class);
+        $container->expects($this->exactly(3))
+            ->method('has')
+            ->withConsecutive(['config'], ['doctrine.connection.orm_foo'], ['doctrine.configuration.orm_bar'])
+            ->willReturnOnConsecutiveCalls(true, true, true);
+        $container->expects($this->exactly(3))
+            ->method('get')
+            ->withConsecutive(['config'], ['doctrine.connection.orm_foo'], ['doctrine.configuration.orm_bar'])
+            ->willReturnOnConsecutiveCalls($config, $connection, $configuration);
 
         $factory       = new EntityManagerFactory();
-        $entityManager = $factory($container->reveal());
+        $entityManager = $factory($container);
 
         $this->assertSame($connection, $entityManager->getConnection());
         $this->assertSame($configuration, $entityManager->getConfiguration());
     }
 
+    /**
+     * @psalm-return Connection&\PHPUnit\Framework\MockObject\MockObject
+     */
     private function buildConnection() : Connection
     {
-        $connection = $this->prophesize(Connection::class);
-        $connection->getEventManager()->willReturn($this->prophesize(EventManager::class)->reveal());
-
-        return $connection->reveal();
+        $eventManager = $this->createStub(EventManager::class);
+        $connection   = $this->getMockBuilder(Connection::class)
```

Comment:
I think the static analysis issue comes from using the mock builder. Psalm can't easily determine the type as `Connection&MockObject` and `getMock` just returns `MockObject`. Can you use `$this->createPartialMock(Connection::class, ['getEventManager']);` perhaps?

Reply:
Yup. You were right. All issues are gone now with b83e443be52a7af5285ce3d74e71fd1d2e07695d.